charms/kfp-ui/src/charm.py

```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python3
+
+import json
+import logging
+
+from ops.charm import CharmBase
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    Unit,
+    WaitingStatus,
+)
+
+from oci_image import OCIImageResource, OCIImageResourceError
+from serialized_data_interface import (
+    NoCompatibleVersions,
+    NoVersionsListed,
+    get_interfaces,
+)
+
+log = logging.getLogger()
+
+
+class Operator(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not self.model.unit.is_leader():
+            log.info("Not a leader, skipping set_pod_spec")
+            self.model.unit.status = ActiveStatus()
+            return
+
+        self.image = OCIImageResource(self, "oci-image")
+
+        try:
+            self.interfaces = get_interfaces(self)
+        except NoVersionsListed as err:
+            self.model.unit.status = WaitingStatus(str(err))
+            return
+        except NoCompatibleVersions as err:
+            self.model.unit.status = BlockedStatus(str(err))
+            return
+
+        relations = [
+            self.on.install,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+            self.on["object-storage"].relation_changed,
+            self.on["kfp-api"].relation_changed,
+        ]
+
+        for relation in relations:
+            self.framework.observe(relation, self.set_pod_spec)
+
+        provides = [
+            self.on["kfp-api"].relation_joined,
+            self.on["kfp-api"].relation_changed,
+        ]
+        for relation in provides:
+            self.framework.observe(relation, self.send_info)
+
+    def send_info(self, event):
+        event.relation.data[self.unit]["service"] = self.model.app.name
+        event.relation.data[self.unit]["port"] = self.model.config["http-port"]
+
+    def set_pod_spec(self, event):
+        try:
+            image_details = self.image.fetch()
+        except OCIImageResourceError as e:
+            self.model.unit.status = e.status
+            log.info(e)
+            return
+
+        config = self.model.config
+
+        if not ((os := self.interfaces["object-storage"]) and os.get_data()):
+            self.model.unit.status = WaitingStatus(
+                "Waiting for object-storage relation data"
+            )
+            return
+        os = list(os.get_data().values())[0]
+
+        try:
+            data = next(
+                value
+                for relation in self.model.relations["kfp-api"]
+                for key, value in relation.data.items()
+                if isinstance(key, Unit)
+                and not key._is_our_unit
+                and value.get("service")
+            )
+            api = {
+                "service": data["service"],
+                "port": data["port"],
+            }
```

Comment:
This could be done differently by using the k8s-service interface.

Reply:
Upgraded to reference https://github.com/canonical/operator-schemas/blob/master/k8s-service.yaml